teps/0121-unified-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,152 @@
+1. When to retry
+2. The amount of attempts
+3. Actions to take after a failed attempt
+4. Timeout of each attempt
+
+
+|  | [Retry Step in GA](https://github.com/marketplace/actions/retry-step) | [Retry Action in GA](https://github.com/marketplace/actions/retry-action) | [Argo Workflow](https://argoproj.github.io/argo-workflows/retries/#retry-policies) | [GitLab](https://docs.gitlab.com/ee/ci/yaml/#retry) | [Ansible](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html#retrying-a-task-until-a-condition-is-met)| [Concourse](https://concourse-ci.org/attempts-step.html#attempts-step) |
+|:---|:---|:---|:---|:---|:---|:---|
+| **When to Retry** | configurable | on failure |configurable|configurable|[always retry, conditional stop](https://github.com/ansible/ansible/pull/76101) [^ansible-conditional-stop]|configurable|
+| **Attempts amount** |supported|supported|supported|supported|supported|supported|
+| **Timeout for each attempt** |supported|supported|supported|[supported](https://docs.gitlab.com/ee/ci/yaml/#retrywhen)|supported|supported|
+| **Timeout for all attempts** |[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[?](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout)|-|-|-|
+
+
+Several observations regarding of the feature table above:
+- We can configure **timeout duration per attempt** in all CI systems that support the `retry` functionality.
+- GitHub Action doesn't support retry natively, but because the flexibility of **customized actions**, some users write their own `retry` action to make it work, and those customized actions even support what to do before retrying a failed attempt.
+- Concourse mentioned the reason it [retries per attempt is somewhat arbitrary](https://concourse-ci.org/attempts-step.html#attempts-step).
+- Argo Workflow supports timeout per attempt, but [this stackoverflow question](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout) implies the need of "timeout for all retries".
+
+## Proposal
+We propose to support `retries` in TaskRuns, and solidify that `Timeout` is for each Retry attempt in CustomRun by updating [TEP-0069](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md) and the Developer guidance [docs/runs.md](https://github.com/tektoncd/pipeline/blob/0dd9f9719177054bdf60928ce9b4aab4b85af5bd/docs/runs.md). 
+
+### Notes and Caveats
+One caveat is that **it's hard for PipelineRun Controller to know whether retries is implemented correctly by customers**. If customers want to use a specific CustomRun, they need to check (the documentation or even the code) whether the functionality is supported properly. 
```

Comment:
If we decide to go with the proposed option, I think we should strongly consider removing retriesStatus from CustomRun status (while still keeping retries in the spec). this will allow customRun controllers to implement retries however they want and prevent the pipelinerun controller from making assumptions about how retries are implemented.

Reply:
Similarly, I think this discussion is also related to [this comment thread](https://github.com/tektoncd/community/pull/816#discussion_r976907440) because IMO we need the `retriesStatus` in pipelineController to decide whether the retries is exhausted, let's move our discussions there.

Reply:
sorry to continue this thread-- I think this is actually a separate point. My idea here was that the pipelinerun controller can determine whether a run is running, failed, or succeeded based on its status.conditions. If we want to hand retries responsibility entirely off to run controllers to implement however they would like, why require them to update status.retriesstatus in a certain way? The run controller can keep track of retries in status.extrafields.

Reply:
No problem at all. I'm wondering how does pipelinerun controller determine the run finishes: like what if it failed (the condition is marked as failed), and would be retried in the next reconciliation loop?

Reply:
I think we'd have to clarify our run controller guidance not to update the ConditionSucceeded until the run is finished running (i.e. most runs would probably implement retries differently than how they're currently implemented on taskrun, but that's OK).

Reply:
Indeed. @pritidesai @abayer I'd like to hear more if we should keep the `retriesStatus` because it constraints the implementation details for CustomRun here. As @lbernick  suggested, we can guide Customers to update `ConditionSucceeded` only after retries is exhausted.

Reply:
> If we want to hand retries responsibility entirely off to run controllers to implement however they would like, why require them to update status.retriesstatus in a certain way? The run controller can keep track of retries in status.extrafields.

`status.retriesStatus` is a natural field to store status of each execution. It is part of an implicit contract between tekton controller and custom controller. `status.extrafields` is a very generic and not ideal for storing  such key information.

What is the motivation behind this request?

Reply:
I made this request, because I think we want to minimize the implicit contract between the tekton controller and custom controller as much as possible. Part of the motivation for adding retries to custom tasks is to allow them to implement retries in whatever way makes the most sense, and we take away some of that flexibility if we require them to update retries status in a certain way.

Reply:
Thanks for the detailed comment.

For the advantages of the your poposed overall setup:

> Users have a consistent interface for retries

- For RuntimeObject driven retries, the interface is also consistent -- we pass PipelineTask.retry to the `retries` field in child runtime objects.

> Custom task controller developers get a default implementation of retries for free (by embedding in a pipeline)

- This harms the flexibility of Custom Task -- what if a custom task controller wants to have different retry stragety/policy, instead of just recreating on failure. Most CI/CD systems have the configurable retry stragety. See the [related work](https://github.com/tektoncd/community/blob/6d02bac0ede7758519d4f36e82d835b7252e63fe/teps/0121-unified-retries-for-taskruns-and-customruns.md#related-work).

> "Pipelines in pipeline" can be retried the same as the other resources

- This is an area that I'm not quite sure what needs to be done, and
-  I'm not sure if this discussion should be part of this TEP.

> No changes to the PipelineRun API (not in the spec at least)

- For RuntimeObject driven retries, no changes to the PipelineRun API neither.

> No changes to the TaskRun API (not in the spec at least)

- For RuntimeObject driven retries, indeed, it adds an extra API field `retries`, but this would be an addictive change rather than backward-incompatible change, wdyt?

I believe we should leave the retry implementation to the CustomRun controller, pipelinerun controller shouldn't touch it at all -- it shouldn't rely on the retriesStatus to decide whether the CustomRun is finished or not etc. 

I totally agree that we can get rid of retriesStatus for both CustomRun and TaskRun. For TaskRun, instead of updating retriesStatus, we can create a new TaskRun in taskrun controller for each retry attempt.

Therefore, having TaskRun controller implements the retry can have most of the advantages you mentiond plus
- taskrun retry on its own
- we don't need to deprecate `retries` in Run

Reply:
Thanks for all of the great discussion in the WG meeting Andrea @afrittoli!

Yeah as you mentioned, it doesn't neccesarrily harm the flexibility of CustomTask. The more accurate way to state it is it harms the UX: say a CustomRun controller implemented its own retry strategy, for the users of the Custom Task, they would be confused about which `retries` field to use (`PipelineTask.retries` or `PipelineTask.TaskSpec.SpecializedRetries`): 

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: pr-custom-task-
spec:
  pipelineSpec:
    tasks:
    - name: wait
      timeout: "1s"
      retries: 1 // The common retries field in the PipelineTask
      taskSpec:
        specialized-retries: 5 // Specialized retries field in Custom Task Spec.
        other-spec-fields: foobar
```

Besides, would you mind pointing out which field we need to update in PipelineRun API spec?